designs/sam_build_cmd.rst

```diff
@@ -0,0 +1,440 @@
+    MyFunction2:
+        Type: AWS::Serverless::Function
+        Properties:
+            ...
+            Code: ./source-code2
+            ...
+
+
+To build, package and deploy this app, customers would do the following:
+
+**1. Build:** Run the following command to build all functions in the template and output a SAM template that can be run through
+the package command:
+
+.. code-block:: bash
+
+    # Build the code and write artifacts to ./build folder
+    $ sam build -t template.yaml -b ./build -o built-template.yaml
+
+
+Output of the *sam build* command is a SAM template where CodeUri is pointing to the built artifacts. Note the values of
+Code properties in following output:
+
+.. code-block:: bash
+
+    $ cat built-template.yaml
+    MyFunction1:
+        Type: AWS::Lambda::Function
+        Properties:
+            ...
+            Code: ./build/MyFunction1
+            ...
+
+    MyFunction2:
+        Type: AWS::Serverless::Function
+        Properties:
+            ...
+            CodeUri: ./build/MyFunction2
+            ...
+
+**2. Package and Deploy:** Package the built artifacts by running the *package* command on the template output
+by *build* command
+
+.. code-block:: bash
+
+    # Package the code
+    $ sam package --template-file built-template.yaml --s3-bucket mybucket --output-template-file packaged-template.yaml
+
+    # Deploy the app
+    $ sam deploy --template-file packaged-template.yaml --stack-name mystack
+
+Other Usecases
+~~~~~~~~~~~~~~~
+
+#. **Build Native Dependencies**: Pass the ``--native`` flag to the *build* command
+#. **Out-of-Source Builds**: In this scenario, Lambda function code is present in a folder outside the folder containing
+   the SAM template. Absolute path to these folders are determined at runtime in a build machine. Set the
+   ``--root=/my/folder`` flag to absolute path to the folder relative to which we will resolve relative *CodeUri* paths.
+#. **Inherited dependency manifest**: By default, we will look for a dependency manifest (ex: package.json) at same
+   folder containing SAM template. If a ``--root`` flag is set, we will look for manifest at this folder. If neither
+   locations have a manifest, we will look for a manifest within the folder containing function code. Manifest present
+   within the code folder always overrides manifest at the root.
+#. **Arbitrary build commands**: Override build commands per-runtime by specifying full path to the command in
+   ``.samrc``.
+#. **Build & Run Locally**: Use the ``--template`` property of ``sam local`` suite of commands to specify the
+   template produced by *build* command (ex: ``build-template.yaml``)
+
+Implementation
+==============
+
+CLI Changes
+-----------
+*Explain the changes to command line interface, including adding new commands, modifying arguments etc*
+
+#. Adding a new top-level command called ``sam build``.
+#. Add ``built-template.yaml`` to list of default template names searched by ``sam local`` commands
```

Comment:
As we're adding defaults to `sam local` commands, would it be possible to add that as part of `sam package` and `sam deploy` too?

Reply:
Yes we can, but I want to do it separately